From f83af24f656adf6bcda72ba1c42882035beec4eb Mon Sep 17 00:00:00 2001
From: LeoLiu-oc <LeoLiu-oc@zhaoxin.com>
Date: Wed, 23 Feb 2022 16:53:23 +0800
Subject: [PATCH] ata: Add support for disabling PhyRdy Change Interrupt based
 on actual LPM capability
Patch-mainline: Not yet, from openEuler
References: bsn#22
openEuler-commit: f83af24f656adf6bcda72ba1c42882035beec4eb
Modified-by-SEL: No


zhaoxin inclusion
category: bugfix
bugzilla: https://gitee.com/openeuler/kernel/issues/I40QDN
CVE: NA

----------------------------------------------------------------

The ahci spec mentions that PhyRdy Change Interrupt and Link Power
Management (LPM) do not coexist. However, before enabling LPM,
the driver did not check whether the host supports LPM,
but directly disabled PhyRdy Change Interrupt. Increase the
judgment on the actual support of LPM, and disable PhyRdy Change
Interrupt only when it is supported.

Signed-off-by: LeoLiu-oc <LeoLiu-oc@zhaoxin.com>
Signed-off-by: Zheng Zengkai <zhengzengkai@huawei.com>
Acked-by: Jackie Liu <liuyun01@kylinos.cn>
Reviewed-by: Jason Yan <yanaijie@huawei.com>
Signed-off-by: Guoqing Jiang <guoqing.jiang@suse.com>
---
 drivers/ata/ahci.c      | 11 +++++++++++
 drivers/ata/libata-eh.c |  7 +++++++
 include/linux/libata.h  |  4 ++++
 3 files changed, 22 insertions(+)

diff --git a/drivers/ata/ahci.c b/drivers/ata/ahci.c
index 8a7140ee88b5..abc2df4a96be 100644
--- a/drivers/ata/ahci.c
+++ b/drivers/ata/ahci.c
@@ -1881,6 +1881,17 @@ static int ahci_init_one(struct pci_dev *pdev, const struct pci_device_id *ent)
 	else
 		dev_info(&pdev->dev, "SSS flag set, parallel bus scan disabled\n");
 
+	if (pdev->vendor == PCI_VENDOR_ID_ZHAOXIN) {
+		if (hpriv->cap & HOST_CAP_PART)
+			host->flags |= ATA_HOST_PART;
+
+		if (hpriv->cap & HOST_CAP_SSC)
+			host->flags |= ATA_HOST_SSC;
+
+		if (hpriv->cap2 & HOST_CAP2_SDS)
+			host->flags |= ATA_HOST_DEVSLP;
+	}
+
 	if (pi.flags & ATA_FLAG_EM)
 		ahci_reset_em(host);
 
diff --git a/drivers/ata/libata-eh.c b/drivers/ata/libata-eh.c
index 018ed8736a64..271c3c76f4af 100644
--- a/drivers/ata/libata-eh.c
+++ b/drivers/ata/libata-eh.c
@@ -3237,6 +3237,8 @@ static int ata_eh_set_lpm(struct ata_link *link, enum ata_lpm_policy policy,
 			  struct ata_device **r_failed_dev)
 {
 	struct ata_port *ap = ata_is_host_link(link) ? link->ap : NULL;
+	struct device *device = ap->host->dev;
+	struct pci_dev *pdev = (!device || !dev_is_pci(device)) ? NULL : to_pci_dev(device);
 	struct ata_eh_context *ehc = &link->eh_context;
 	struct ata_device *dev, *link_dev = NULL, *lpm_dev = NULL;
 	enum ata_lpm_policy old_policy = link->lpm_policy;
@@ -3245,6 +3247,11 @@ static int ata_eh_set_lpm(struct ata_link *link, enum ata_lpm_policy policy,
 	unsigned int err_mask;
 	int rc;
 
+	/* if controller does not support lpm, then sets no LPM flags*/
+	if ((pdev && pdev->vendor == PCI_VENDOR_ID_ZHAOXIN) &&
+		!(ap->host->flags & (ATA_HOST_PART | ATA_HOST_SSC | ATA_HOST_DEVSLP)))
+		link->flags |= ATA_LFLAG_NO_LPM;
+
 	/* if the link or host doesn't do LPM, noop */
 	if (!IS_ENABLED(CONFIG_SATA_HOST) ||
 	    (link->flags & ATA_LFLAG_NO_LPM) || (ap && !ap->ops->set_lpm))
diff --git a/include/linux/libata.h b/include/linux/libata.h
index d3600fc7f7c4..bbce7a124286 100644
--- a/include/linux/libata.h
+++ b/include/linux/libata.h
@@ -260,6 +260,10 @@ enum {
 	ATA_HOST_PARALLEL_SCAN	= (1 << 2),	/* Ports on this host can be scanned in parallel */
 	ATA_HOST_IGNORE_ATA	= (1 << 3),	/* Ignore ATA devices on this host. */
 
+	ATA_HOST_PART		= (1 << 4), /* Host support partial.*/
+	ATA_HOST_SSC		= (1 << 5), /* Host support slumber.*/
+	ATA_HOST_DEVSLP		= (1 << 6), /* Host support devslp.*/
+
 	/* bits 24:31 of host->flags are reserved for LLD specific flags */
 
 	/* various lengths of time */
-- 
2.26.2

